areas/mysql/src/AzureMcp.MySql/Services/MySqlService.cs

```diff
@@ -0,0 +1,481 @@
+    }
+
+    private async Task<string> BuildConnectionStringAsync(string server, string user, string database)
+    {
+        var entraIdAccessToken = await GetEntraIdAccessTokenAsync();
+        var host = NormalizeServerName(server);
+        return $"Server={host};Database={database};User ID={user};Password={entraIdAccessToken};SSL Mode=Required;";
+    }
+
+    private static void ValidateQuerySafety(string query)
+    {
+        if (string.IsNullOrWhiteSpace(query))
+        {
+            throw new ArgumentException("Query cannot be null or empty.", nameof(query));
+        }
+
+        // Prevent DoS attacks by limiting query length
+        if (query.Length > 10000)
+        {
+            throw new InvalidOperationException("Query length exceeds the maximum allowed limit of 10,000 characters to prevent potential DoS attacks.");
+        }
+
+        // Clean the query: remove comments, normalize whitespace, and trim
+        var cleanedQuery = query;
+
+        // Remove line comments (-- comment)
+        cleanedQuery = Regex.Replace(cleanedQuery, @"--.*?$", "", RegexOptions.Multiline);
+
+        // Remove hash comments (# comment)
+        cleanedQuery = Regex.Replace(cleanedQuery, @"#.*?$", "", RegexOptions.Multiline);
+
+        // Remove block comments (/* comment */)
+        cleanedQuery = Regex.Replace(cleanedQuery, @"/\*.*?\*/", "", RegexOptions.Singleline);
+
+        // Normalize whitespace: replace multiple whitespace characters with single space
+        cleanedQuery = Regex.Replace(cleanedQuery, @"\s+", " ", RegexOptions.Multiline);
+
+        // Trim the result
+        cleanedQuery = cleanedQuery.Trim();
+
+        // Ensure the cleaned query is not empty
+        if (string.IsNullOrWhiteSpace(cleanedQuery))
+        {
+            throw new ArgumentException("Query cannot be empty after removing comments and whitespace.", nameof(query));
+        }
+
+        // Check for multiple SQL statements (semicolons that are not in comments)
+        var semicolonCount = cleanedQuery.Split(';').Length - 1;
+        if (semicolonCount > 1)
+        {
+            throw new InvalidOperationException("Multiple SQL statements are not allowed. Use only a single SELECT statement.");
+        }
+
+        // Regex pattern to detect common SQL injection techniques (comments are already removed)
+        var dangerPattern = new Regex(
+            @"(;|\b(union\s+select|drop\s+table|insert\s+into|update\s+|delete\s+from|or\s+1=1)\b)",
+            RegexOptions.IgnoreCase | RegexOptions.Compiled
+        );
+
+        // Check for common SQL injection patterns using regex
+        if (dangerPattern.IsMatch(cleanedQuery))
+        {
+            throw new InvalidOperationException("Query contains dangerous patterns that could indicate SQL injection attempts and is not allowed for security reasons.");
+        }
+
+        // List of dangerous SQL keywords that should be blocked
+        var dangerousKeywords = new[]
```

Comment:
Does this array really need to be recreated each time a query is being sanitized?

Reply:
pushed to private